### PR TITLE
feat(meta): support replace source

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Finish the meta part of #19063: 
- 1. extend `match &stream_job` from Table to Source
- 2. handle split assignment for replace (need to align to prev actors, because of source backfill is aligned)
- Also changed the grpc interface from table to job.

e2e tests can be seem in #19649

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19650?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#19649** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19649?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#19650** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19650?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19650?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18935** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18935?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#19630** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19630?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/meta/src/stream/source_manager.rs [hunk 1]
```diff
@@ -972,13 +1011,89 @@ impl SourceManager {
...
+        // old_source           backfill_1            new_source
+        // actor_x1-----┬------>actor_a1 -----┬------>actor_y1
+        // actor_x2-----┼-┬---->actor_a2 -----┼-┬---->actor_y2
+        //              │ │                   │ │
+        //              │ │     backfill_2    │ │
+        //              └-┼---->actor_b1------┘ │
+        //                └---->actor_b2--------┘
+        //
+        // Note: we can choose any backfill actor to align here.
+        // We use `HashMap` to dedup.
```
Comment (chenzl25) [inline src/meta/src/stream/source_manager.rs:1067]: I don't quite get why backfill is appearing here.

**Review by chenzl25 — APPROVED** [review]
Generally LGTM.

File: src/meta/src/stream/source_manager.rs (hunk 1 above)
Comment (xxchan) [reply src/meta/src/stream/source_manager.rs:1067]: - When creating SourceBackfill executors, we need to align split assignment to the source executor. (i.e., actors connected with NoShuffle should have the same splits)
- Therefore, the real goal is that when replacing source executors, we need to ensure the new executors still have the aligned splits.

File: src/meta/src/stream/source_manager.rs [hunk 2]
```diff
@@ -495,6 +497,32 @@ impl SourceManagerCore {
                 self.actor_splits.remove(actor_id);
             }
         }
+
+        if let Some(fragment_replacements) = fragment_replacements {
+            for (old_fragment_id, new_fragment_id) in fragment_replacements {
+                for fragment_ids in self.source_fragments.values_mut() {
+                    if fragment_ids.remove(&old_fragment_id) {
+                        fragment_ids.insert(new_fragment_id);
+                    }
```
Comment (BugenZhao) [inline src/meta/src/stream/source_manager.rs:506]: Why didn't we have this step prior to this PR, considering we've already supported schema change on table with source?

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM

File: src/meta/src/model/stream.rs [hunk 3]
```diff
@@ -369,6 +369,10 @@ impl StreamJobFragments {
         actor_ids
     }
 
+    pub fn root_fragment(&self) -> Option<Fragment> {
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/meta/src/model/stream.rs:372]: Do you have suggestions?

File: src/meta/src/stream/source_manager.rs (hunk 1 above)
Comment (xxchan) [reply src/meta/src/stream/source_manager.rs:1067]: The graph might be a little misleading.

File: src/meta/src/stream/source_manager.rs (hunk 2 above)
Comment (xxchan) [reply src/meta/src/stream/source_manager.rs:506]: I had the same question. I think it's overlooked, i.e., we forgot to GC old fragment here. New fragment will be added by `added_source_fragments` above.

File: proto/ddl_service.proto [hunk 4]
```diff
@@ -353,26 +353,42 @@ message DropIndexResponse {
   WaitVersion version = 2;
 }
 
-message ReplaceTablePlan {
-  // The new table catalog, with the correct (old) table ID and a new version.
-  // If the new version does not match the subsequent version in the meta service's
-  // catalog, this request will be rejected.
-  catalog.Table table = 1;
+message ReplaceJobPlan {
+  reserved 1;
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply proto/ddl_service.proto:357]: Yes

**Comment by xxchan** [comment]
@BugenZhao look look

**Review by BugenZhao — APPROVED** [review]

**Review by yezizp2012 — APPROVED** [review]

**Pull request merged**